File: Sorting/Merge_Sort.cpp
// TIME COMPLEXITY : O(nlogn)
// SPACE COMPLEXITY : O(n)

/*
Basic Idea:
Divide And Conquer Method (Recursion)
*/

/*
APPLICATION:
1. sorting linked list
2. Inversion count problems
3. Finding the median of an array
*/

#include <iostream>
using namespace std;

void merge(int *arr, int start, int end)
{
    int mid = start + (end - start) / 2;

    // LENGTH OF TWO SEPRATED ARRAY
    int length1 = mid - start + 1;
    int length2 = end - mid;

    // COPY THE TWO SEPRATED ARRAY
    int *first = new int[length1];
    int *second = new int[length2];

    // COPY VALUES TO ARRAY 1
    int mainArrayIndex = start;
    for (int i = 0; i < length1; i++)
    {
        first[i] = arr[mainArrayIndex++];
    }

    // COPY VALUES TO ARRAY 2
    mainArrayIndex = mid + 1;
    for (int i = 0; i < length2; i++)
    {
        second[i] = arr[mainArrayIndex++];
    }

    // MERGE TWO SORTED ARRAY
    int index1 = 0;
    int index2 = 0;
    mainArrayIndex = start;

    while (index1 < length1 && index2 < length2)
    {
        if (first[index1] < second[index2])
        {
            arr[mainArrayIndex++] = first[index1++];
        }
        else
        {
            arr[mainArrayIndex++] = second[index2++];
        }
    }

    while (index1 < length1)
    {
        arr[mainArrayIndex++] = first[index1++];
    }

    while (index2 < length2)
    {
        arr[mainArrayIndex++] = second[index2++];
    }

    delete[] first;
    delete[] second;
}

void mergeSort(int *arr, int start, int end)
{
    // BASE CASE
    if (start >= end)
        return;

    int mid = start + (end - start) / 2;

    mergeSort(arr, start, mid);
    mergeSort(arr, mid + 1, end);

    merge(arr, start, end);
}

int main()
{
    int arr[5] = {6, 2, 8, 4, 10};
    mergeSort(arr, 0, 5);

    for (int i = 0; i < 5; i++)
        cout << arr[i] << " ";

    return 0;
}

File: Sorting/MergeSort.js
/**
 * Implements Merge Sort algorithm
 * @param {number[]} arr - Array to be sorted
 * @returns {number[]} - Sorted array
 */
function mergeSort(arr) {
    if (arr.length <= 1) return arr;
    
    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));
    
    return merge(left, right);
}

/**
 * Merges two sorted arrays
 * @param {number[]} left - First sorted array
 * @param {number[]} right - Second sorted array
 * @returns {number[]} - Merged sorted array
 */
function merge(left, right) {
    const result = [];
    let i = 0, j = 0;
    
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
    }
    
    return result.concat(left.slice(i)).concat(right.slice(j));
}

// Example usage
const arr = [64, 34, 25, 12, 22, 11, 90];
console.log("Original array:", arr);
console.log("Sorted array:", mergeSort(arr)); 